### PR TITLE
No longer require string what/with

### PR DESCRIPTION
**Pull request by hadley**
Turned out to be very simple after all the refactoring 😄

### CODE REVIEW

File: R/deprecated.R [hunk 1]
```diff
@@ -58,32 +58,32 @@
...
+#' deprecate_warn("1.0.0", foo(arg = 'must be a scalar integer'))
 #'
 #' # A deprecated function with a function replacement:
-#' deprecate_warn("1.0.0", "foo()", "bar()")
+#' deprecate_warn("1.0.0", foo(), bar())
 #'
 #' # A deprecated function with a function replacement from a
 #' # different package:
-#' deprecate_warn("1.0.0", "foo()", "otherpackage::bar()")
+#' deprecate_warn("1.0.0", foo(), "otherpackage::bar()")
```
Comment (lionel-) [inline R/deprecated.R:74]: Still quoted: `"otherpackage::bar()"`

File: R/spec.R [hunk 2]
```diff
@@ -21,12 +21,14 @@ spec <- function(spec, env = caller_env(), signaller = "signal_lifecycle") {
...
 
 spec_what <- function(what, arg, signaller) {
-  if (!is_string(what)) {
-    lifecycle_abort("`what` must be a string")
+  if (is_call(what)) {
+    call <- what
+  } else if (is_string(what)) {
+    call <- parse_expr(what)
+  } else {
+    lifecycle_abort("`what` must be a string or call")
```
Comment (lionel-) [inline R/spec.R:29]: Missing full stop.

```suggestion
    lifecycle_abort("`what` must be a string or call.")
```

File: R/spec.R [hunk 3]
```diff
@@ -21,12 +21,14 @@ spec <- function(spec, env = caller_env(), signaller = "signal_lifecycle") {
 }
 
 spec_what <- function(what, arg, signaller) {
-  if (!is_string(what)) {
-    lifecycle_abort("`what` must be a string")
+  if (is_call(what)) {
+    call <- what
+  } else if (is_string(what)) {
+    call <- parse_expr(what)
```
Comment (lionel-) [inline R/spec.R:27]: Do we dare soft-deprecating this? 🙃

File: tests/testthat/test-deprecated.R [hunk 4]
```diff
@@ -4,35 +4,35 @@ test_that("default deprecations behave as expected", {
   on.exit(env_unbind(deprecation_env, "test"))
   local_options(lifecycle_verbosity = "default")
 
-  expect_condition(deprecate_soft("1.0.0", "foo()"), class = "lifecycle_soft_deprecated")
+  expect_condition(deprecate_soft("1.0.0", foo()), class = "lifecycle_soft_deprecated")
   expect_warning(deprecate_warn("1.0.0", "foo()", id = "test"), class = "lifecycle_warning_deprecated")
```
Comment (lionel-) [inline tests/testthat/test-deprecated.R:8]:
```suggestion
  expect_warning(deprecate_warn("1.0.0", foo(), id = "test"), class = "lifecycle_warning_deprecated")
```

**Review by lionel- — APPROVED** [review]
Cool!

File: R/deprecated.R (hunk 1 above)
Comment (hadley) [reply R/deprecated.R:74]: Yeah, otherwise R CMD check warns about use of `otherpackage`

File: R/spec.R (hunk 3 above)
Comment (hadley) [reply R/spec.R:27]: It's tempting, but I think it will cause more hassle than its worth.

**Pull request merged**